core/src/main/java/io/questdb/cairo/sql/ArgFunctions.java

```diff
@@ -0,0 +1,332 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.sql;
+
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.BinaryFunction;
+import io.questdb.griffin.engine.functions.TernaryFunction;
+import io.questdb.griffin.engine.functions.UnaryFunction;
+import io.questdb.griffin.engine.functions.constants.StrConstant;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Utility class for argument {@link Function}s.
+ */
+public class ArgFunctions {
```

Comment:
We try to avoid allocations as much as possible, even on non-hot path. This utility class assumes allocations that could be avoided, so I'd prefer to see an approach similar to what we have in other factories.

@bluestreak01 WDYT?

Reply:
Ok, I'll rework it to use the existing approach, in hindsight I should have discussed about it before fully implementing it.

I'll create a separate PR for the abstraction in case you all ever change your mind

Reply:
Thanks you so much!

> I'll create a separate PR for the abstraction in case you all ever change your mind

Yeah, why not. We're definitely open to any improvements.

Reply:
LGTM. Sorry for the delays in review iterations.

Reply:
@puzpuzpuz I've created a new PR for the abstraction at #2679

Reply:
Thanks. I've noticed it. We're pretty busy with the upcoming replication feature, but someone from the team should be able to review it soon.